src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs

```diff
@@ -224,55 +227,66 @@ private void SendFileInternal(string? fileName, byte[]? preBuffer, byte[]? postB
             if (errorCode != SocketError.Success)
             {
                 UpdateSendSocketErrorForDisposed(ref errorCode);
-
                 UpdateStatusAfterSocketErrorAndThrowException(errorCode);
             }
 
             // Send the postBuffer, if any
             // This will throw on error
-            if (postBuffer != null && postBuffer.Length > 0)
+            if (!postBuffer.IsEmpty)
             {
                 Send(postBuffer);
             }
         }
 
-        private async Task SendFileInternalAsync(FileStream? fileStream, byte[]? preBuffer, byte[]? postBuffer)
+        private async ValueTask SendFileInternalAsync(FileStream? fileStream, ReadOnlyMemory<byte> preBuffer, ReadOnlyMemory<byte> postBuffer, TransmitFileOptions flags = TransmitFileOptions.UseDefaultWorkerThread, CancellationToken cancellationToken = default)
         {
+            CheckTransmitFileOptions(flags);
+
             SocketError errorCode = SocketError.Success;
             using (fileStream)
             {
                 // Send the preBuffer, if any
                 // This will throw on error
-                if (preBuffer != null && preBuffer.Length > 0)
+                if (!preBuffer.IsEmpty)
                 {
                     // Using "this." makes the extension method kick in
-                    await this.SendAsync(new ArraySegment<byte>(preBuffer), SocketFlags.None).ConfigureAwait(false);
+                    await SendAsync(preBuffer, SocketFlags.None, cancellationToken).ConfigureAwait(false);
                 }
 
                 // Send the file, if any
                 if (fileStream != null)
                 {
-                    var tcs = new TaskCompletionSource<SocketError>();
-                    errorCode = SocketPal.SendFileAsync(_handle, fileStream, (_, socketError) => tcs.SetResult(socketError));
+                    AsyncValueTaskMethodBuilder<SocketError> taskBuilder = AsyncValueTaskMethodBuilder<SocketError>.Create();
+                    ValueTask<SocketError> task = taskBuilder.Task;
+                    errorCode = SocketPal.SendFileAsync(_handle, fileStream, taskBuilder, static (state, _, socketError) => state.SetResult(socketError), cancellationToken);
                     if (errorCode == SocketError.IOPending)
                     {
-                        errorCode = await tcs.Task.ConfigureAwait(false);
+                        errorCode = await task.ConfigureAwait(false);
                     }
                 }
             }
 
             if (errorCode != SocketError.Success)
             {
                 UpdateSendSocketErrorForDisposed(ref errorCode);
-                UpdateStatusAfterSocketErrorAndThrowException(errorCode);
+
+                if (errorCode == SocketError.OperationAborted)
```

Comment:
Instead of checking for OperationAborted, I think we should just check if the cancellationToken is cancelled and throw if it is (i.e. call cancellationToken.ThrowIfCancellationRequested)

Reply:
Though we may want to call UpdateStatusAfterSocketError too, not sure.

Reply:
> we may want to call UpdateStatusAfterSocketError too, not sure.

I'm not sure either, but I think we won't. https://github.com/dotnet/runtime/blob/0262b49a470bff6353fc91e87aec18b2c13c9e69/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L4022-L4036 would disconnect the socket, and cancellation is no reason to disconnect, or?